src/SpeedyTransforms/spectral_transform.jl

```diff
@@ -340,38 +346,33 @@ function transform!(                # SPECTRAL TO GRID
     unscale_coslat::Bool=false,     # unscale with cos(lat) on the fly?
 ) where NF                          # number format NF
 
-    (; nlat, nlons, nlat_half, nfreq_max ) = S
-    (; cos_colat, sin_colat, lon_offsets ) = S
-    (; recompute_legendre, Λ, Λs, m_truncs ) = S
-    (; brfft_plans ) = S
+    (; nlat, nlons, nlat_half) = S                  # dimensions
+    (; lmax, mmax ) = S                             # 0-based max degree l, order m of spherical harmonics
+    (; cos_colat, sin_colat, lon_offsets ) = S  
+    (; recompute_legendre, Λ, Λs, m_truncs ) = S    # for Legendre transform
+    (; brfft_plans) = S                             # for Fourier transform
 
     @boundscheck ismatching(S, grids) || throw(DimensionMismatch(S, grids))
     @boundscheck ismatching(S, specs) || throw(DimensionMismatch(S, specs))
+    @boundscheck size(grids, 2) == size(specs, 2) <= S.nlayers || throw(DimensionMismatch(S, specs))
 
-    lmax = specs.m - 1            # 0-based maximum degree l of spherical harmonics
-    mmax = specs.n - 1            # 0-based maximum order m of spherical harmonics
-
-    # preallocate work arrays
-    gn = zeros(Complex{NF}, nfreq_max)      # phase factors for northern latitudes
-    gs = zeros(Complex{NF}, nfreq_max)      # phase factors for southern latitudes
-
-    rings = eachring(grids)                 # precomputed ring indices
+    # use scratch memory for Legendre but not yet Fourier transformed data
+    gn = S.scratch_memory_north                 # phase factors for northern latitudes
+    gs = S.scratch_memory_south                 # phase factors for northern latitudes
 
     Λw = Legendre.Work(Legendre.λlm!, Λ, Legendre.Scalar(zero(Float64)))
 
-    # loop over all specs/grids (e.g. vertical dimension)
-    # k, k_grid only differ when specs/grids have a singleton dimension
-    @inbounds for (k, k_grid) in zip(eachmatrix(specs), eachgrid(grids))
-        for j_north in 1:nlat_half              # symmetry: loop over northern latitudes only
-            j_south = nlat - j_north + 1        # southern latitude index
-            nlon = nlons[j_north]               # number of longitudes on this ring
-            nfreq  = nlon÷2 + 1                 # linear max Fourier frequency wrt to nlon
-            m_trunc = m_truncs[j_north]         # (lin/quad/cub) max frequency to shorten loop over m
-            not_equator = j_north != j_south    # is the latitude ring not on equator?
+    # INVERSE LEGENDRE TRANSFORM
```

Comment:
Just an idea:

To remove code redundancy, and make it a more flexible, could we put the Legende transform in a separate function and let both transforms use the same (with the quadrature weights optional just used by one of them)? 

I am thinking about this because for the pullback/adjoint version, essentially you reverse the operators, so for a Fourier transform the adjoint is the inverse transform because of the complex conjugation. But for the SPH transform, the adjoint of the Legendre transform, doesn't include the quadrature weights, it's just the transpose of the original transform. So, we need some flexibility there for the differentiable version anyway. We can also retrofit this though.

Reply:
Thinking about this again, for the GPU version, we'd probably do different versions for CPU and GPU. In theory KernalAbstractions can run on CPU of course, but it's very unlikely we'll get the KA CPU version to the same speed as this optimised pure CPU version. The transform is by far the most critical component in terms of performance for the model, so it's worth it to keep two versions. We could do this by just having a dispatch within the `transform!` function. So 

```pseudocode  
function transform!(grids, specs, S) 
       checks and preparation 

       LT!(gnorth, gsouth, specs, legendre_pols) 
       FT!(grids, specs, plans) 
end 

function LT!(gnorth::CuArray, ... ) # or dispatch based on Device to allow other non-CUDA GPUs 

end 

function LT!(gnorth::Array,...)

end

Reply:
On it! But there's a few more things that need to be passed on, so currently I have `S::SpectralTransform` as an argument for both `_legendre!` and `_fourier!` 
```julia
function transform!(                    # SPECTRAL TO GRID
    grids::AbstractGridArray,           # gridded output
    specs::LowerTriangularArray,        # spectral coefficients input
    S::SpectralTransform;               # precomputed transform
    unscale_coslat::Bool = false,       # unscale with cos(lat) on the fly?
)
    # use scratch memory for Legendre but not yet Fourier-transformed data
    g_north = S.scratch_memory_north    # phase factors for northern latitudes
    g_south = S.scratch_memory_south    # phase factors for northern latitudes

    # INVERSE LEGENDRE TRANSFORM in meridional direction
    _legendre!(g_north, g_south, specs, S; unscale_coslat)

    # INVERSE FOURIER TRANSFORM in zonal direction
    _fourier!(grids, g_north, g_south, S)

    return grids
end
```

Reply:
Done that now for the inverse already, makes it easier to understand too, I find. Gave the functions some `_` because they are really not meant to be used by itself